### PR TITLE
Add atomic loads and stores

### PR DESCRIPTION
**Pull request by dschuff**
Add IR, wast and binary support for atomic loads and stores.

Currently all IR generated by means other than parsing wast and binary files always generates non-atomic accesses, and optimizations have not yet been made aware of atomics, so they are certainly not ready to be used yet.

### CODE REVIEW

File: src/compiler-support.h [hunk 1]
```diff
@@ -27,7 +27,7 @@
 
 // If control flow reaches the point of the WASM_UNREACHABLE(), the program is
 // undefined.
-#if __has_builtin(__builtin_unreachable)
+#if __has_builtin(__builtin_unreachable) && defined(NDEBUG)
```
Comment (kripken) [inline src/compiler-support.h]: what does this change do?

File: src/wasm/wasm-binary.cpp [hunk 2]
```diff
@@ -1873,13 +1926,19 @@ BinaryConsts::ASTNodes WasmBinaryBuilder::readExpression(Expression*& curr) {
...
+      throw ParseException("invalid code after atomic prefix: " + std::to_string(code));
+    }
     default: {
       // otherwise, the code is a subcode TODO: optimize
       if (maybeVisitBinary(curr, code)) break;
       if (maybeVisitUnary(curr, code)) break;
       if (maybeVisitConst(curr, code)) break;
-      if (maybeVisitLoad(curr, code)) break;
-      if (maybeVisitStore(curr, code)) break;
+      if (maybeVisitLoad(curr, code, /*isAtomc=*/false)) break;
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: typos in `Atomic`

File: src/compiler-support.h (hunk 1 above)
Comment (dschuff) [inline src/compiler-support.h]: It means that it uses `__builtin_unreachable` only in release mode, and aborts in debug mode. Makes catching bugs easier.

**Comment by kripken** [comment]
I think we might need some followup work here:

 * This PR added a property to loads and stores. We need to make hashing, copying and equality checking work with that. This could be a silent error, so it's worrying.
 * The new atomic expression type needs support for those things as well. Also possible that it needs support in some of the optimization passes. This should cause aborts, so it's less worrying.

**Comment by dschuff** [comment]
Yes, we definitely need followup work. I probably should have said more about my current plan in a bug or something. But anyway, yeah this is definitely the beginning of adding atomic/shared mem support.
This PR (and the upcoming ones to add atomicrmw/cmpxchg etc) focus on getting the instructions in, and round-tripping via text and binary format. Then we definitely need to look at optimizations and such. Certainly many optimizations will need to treat atomic loads/stores differently, and atomicrmw and friends will be completely different IR instructions. Concurrently we can add asm2wasm/s2wasm support to actually use them.

**Pull request merged**